src/test/java/com/balako/onlinebookstore/service/BookServiceTest.java

```diff
@@ -0,0 +1,243 @@
+package com.balako.onlinebookstore.service;
+
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyLong;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoInteractions;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
+import static org.mockito.Mockito.when;
+
+import com.balako.onlinebookstore.dto.book.request.BookSearchParametersDto;
+import com.balako.onlinebookstore.dto.book.request.CreateBookRequestDto;
+import com.balako.onlinebookstore.dto.book.response.BookDto;
+import com.balako.onlinebookstore.dto.book.response.BookDtoWithoutCategoryIds;
+import com.balako.onlinebookstore.exception.EntityNotFoundException;
+import com.balako.onlinebookstore.mapper.BookMapper;
+import com.balako.onlinebookstore.model.Book;
+import com.balako.onlinebookstore.model.Category;
+import com.balako.onlinebookstore.repository.book.BookRepository;
+import com.balako.onlinebookstore.repository.book.BookSpecificationBuilder;
+import com.balako.onlinebookstore.service.impl.BookServiceImpl;
+import java.math.BigDecimal;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+import org.springframework.data.domain.PageRequest;
+import org.springframework.data.domain.Pageable;
+import org.springframework.data.jpa.domain.Specification;
+
+@ExtendWith(MockitoExtension.class)
+public class BookServiceTest {
+    private static final Book BOOK = new Book();
+    private static final Category CATEGORY = new Category();
+    private static final CreateBookRequestDto REQUEST_DTO =
+            new CreateBookRequestDto();
+    private static final BookDtoWithoutCategoryIds DTO_WITHOUT_CATEGORY_IDS =
+            new BookDtoWithoutCategoryIds();
+    private static final BookDto BOOK_DTO = new BookDto();
+    private static final Long BOOK_ID = 1L;
+    private static final Long CATEGORY_ID = 1L;
+    private static final Integer NUMBER_OF_INVOCATIONS = 1;
+    private static final BookSearchParametersDto SEARCH_PARAMETERS =
+            new BookSearchParametersDto(
+                    new String[] {"Title 1", "Title 2"},
+                    new String[] {"Author 1", "Author 2"},
+                    new String[] {"978-0545010221"});
```

Comment:
Maybe it's better to use your creativity? But it doesn't matter

Reply:
Creativity?🤔 Know only creative mode in minecraft 🧊😆 
I'll ask you in PM.